### DISCUSSION TITLE
Generating Prisma Client Unable to Read DATABASE_URL in .env | InvalidDatasourceError | Deno

### DISCUSSION THREAD

**Discussion by riefer02** [prisma/prisma#16444] [Q&A] [answered]
Hello Prisma Team,

I am attempting to use Prisma cloud to generate my client for my deno instance after following the documentation [here](https://www.prisma.io/docs/guides/deployment/deployment-guides/deploying-to-deno-deploy).

I have `DATABASE_URL` created by Prisma studio set in the project's top level `.env`

```
DATABASE_URL=prisma://abc-me-north-420.prisma-data.com/?api_key=my-api-key
```

`schema.prisma`:
```
generator client {
  provider        = "prisma-client-js"
  output          = "../generated/client"
  previewFeatures = ["deno"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL") <--- This works if I hard set the value, which leads me to to believe at run time it's not looking at the right spot
}
```

When I run `npx dotenv -- deno run -A --unstable npm:prisma generate --data-proxy`

It seems to build the client successfully.

```
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
napi_add_env_cleanup_hook is currently not supported

✔ Generated Prisma Client (4.6.1 | dataproxy) to ./generated/client in 242ms
You can now start using Prisma Client in your code. Reference: https://pris.ly/d/client

import { PrismaClient } from './generated/client'
const prisma = new PrismaClient()


To use Prisma Client with Deno and the Data Proxy, import it like this:

import { PrismaClient } from './generated/client/deno/edge.ts'


You will need a Prisma Data Proxy connection string. See documentation: https://pris.ly/d/data-proxy

Warning: Not implemented: ChildProcess.prototype.disconnect

```
I now have a `./generated/client` folder that is generated in my project directory.

Here is my start script for the project:

`npx dotenv -- deno run -A --unstable --watch=static/,routes/ dev.ts`

Then when I boot up my deno project I get the following:

```
error: Uncaught (in promise) InvalidDatasourceError: Datasource "db" references an environment variable "DATABASE_URL" that is not set
    at ar.resolveDatasourceURL (file:///Users/andrew/Code/deno-api/generated/client/runtime/edge-esm.js:27:2653)
    at ar.extractHostAndApiKey (file:///Users/andrew/Code/deno-api/generated/client/runtime/edge-esm.js:27:1734)
    at new DataProxyEngine (file:///Users/andrew/Code/deno-api/generated/client/runtime/edge-esm.js:25:2894)
    at t.getEngine (file:///Users/andrew/Code/deno-api/generated/client/runtime/edge-esm.js:90:3468)
    at new PrismaClient (file:///Users/andrew/Code/deno-api/generated/client/runtime/edge-esm.js:90:3026)
    at file:///Users/andrew/Code/deno-api/routes/api/v1/artists/index.tsx:7:16
```

### I Have Tried...

If I directly add the the `DATABASE_URL` to the `schema.prisma` replacing `env("DATABASE_URL")` the client generates successfully and the program executes with no error. 

This all leads me to believe it's just an issue with `schema.prisma` not reading the `DATABASE_URL` at run time.

I have checked the `DATABASE_URL` at runtime by running:

```
$ npx dotenv -p DATABASE_URL

$ prisma://abc-me-north-420.prisma-data.com/?api_key=my-api-key
```

---

I have also tried using the direct `DATABASE_URL` for the PostgreSQL DB rather the Prisma proxy `DATABASE_URL` during the `prisma generate` but still the same error as above.

As of now, the only fix is for me to hardcore the `DATABASE_URL` value but this will not work in a production environment.

Any help is appreciated. Let me know if I can add more clarity and here is the project [repo](https://github.com/riefer02/fresh-deno-api/tree/prisma). I'm going to dive back into the compiled `./generated/client/edge-esm.js` and keep investigating. Cheers.

**Comment by riefer02** [comment, marked as answer]
I just manually fed the `DATABASE_URL` env variable to the Prisma client.

```
import { config } from "https://deno.land/x/dotenv/mod.ts";

const databaseUrl = Deno.env.get("DATABASE_URL")! || config().DATABASE_URL;

import { PrismaClient } from "@prisma/client/edge";
const prisma = new PrismaClient({
  datasources: {
    db: {
      url: databaseUrl,
    },
  },
});

export default prisma;
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
